wordpress-mobile/WordPress-Android
# Enable account closure

### Description

This PR adds support for the user to close their account from the account settings screen.

**Note**: This is a feature branch, and should not be merged until all the respective upstream PRs have been reviewed, tested, and merged onto this branch.

PRs:

* https://github.com/wordpress-mobile/WordPress-Android/pull/18382
* https://github.com/wordpress-mobile/WordPress-Android/pull/18386
* https://github.com/wordpress-mobile/WordPress-Android/pull/18391
* https://github.com/wordpress-mobile/WordPress-Android/pull/18407
* https://github.com/wordpress-mobile/WordPress-Android/pull/18409
* https://github.com/wordpress-mobile/WordPress-Android/pull/18444
* https://github.com/wordpress-mobile/WordPress-Android/pull/18457
* https://github.com/wordpress-mobile/WordPress-Android/pull/18470

(PT: pcdRpT-2yo-p2)


### To test:

**Note**: As this is a feature PR, please test with the latest changes in the HEAD PR's branch: https://github.com/wordpress-mobile/WordPress-Android/pull/18444

<details>
<summary>A note about coverage</summary>

When considering [the possible errors from the API](https://github.com/wordpress-mobile/WordPress-Android/blob/feature/enable-account-closure--18413-add-tracks-events/WordPress/src/main/java/org/wordpress/android/ui/prefs/accountsettings/components/DialogErrorUi.kt#L38-L43), some scenarios may be difficult to test or are not relevant to real users / scenarios. E.g.:

* `UNAUTHORIZED`: This may happen for A8C accounts, but will likely be preempted by the Atomic site error anyway
  * Also, I was not brave enough to try the account closure flow with my main account 😅 .
* `ATOMIC_SITE`: This scenario is covered in the testing steps
* `CHARGEBACKED_SITE`: This may require either mocking the response, or some elaborate manual setup
  * It also doesn't sound like a particularly good idea to test for real 😅 
* `ACTIVE_SUBSCRIPTIONS`: This scenario is covered in the testing steps
* `ACTIVE_MEMBERSHIPS`: This is _not_ tested, mainly because I didn't see an easy way to get a disposable account into this state
* `INVALID_TOKEN`: This state should not be reachable, afaiu (outside of debugging the flow)
* `UNKNOWN`: I don't know how to get this error (no pun intended)

I think the manual testing coverage is reasonable, given that, but please let me know if you feel otherwise.
</details>

<details>
<summary>Prerequisites (helpers)</summary>

In order to test the following flows, for convenience I have prepared some pre-configured accounts to test certain error conditions. Please see ref: pc8HXX-17y-p2 to access those credentials.

I have also prepared some scripts to make generating disposable accounts easier via `adb`
</details>

#### :red_circle: User with an Atomic site

1. Log in with a pre-configured account (with an Atomic site)
2. Navigate to the account settings (Me :arrow_right: Account Settings)
3. Tap "Close Account"
4. Expect that an error dialog is displayed with a button to contact support
   * Expect that the error is tracked: `🔵 Tracked: close_account_failed, Properties: {"error_code":"atomic-site"}`
5. Tap "Contact Support"
6. Expect that the Help screen is displayed
   * Expect that the origin is tracked: `🔵 Tracked: support_opened, Properties: {"origin":"ACCOUNT_CLOSURE_DIALOG"}`
7. Tap back
8. Tap dismiss
9. Expect that the dialog is dismissed

#### :red_circle: User with a Personal site

1. Log in with a pre-configured account (with a Personal site)
2. Navigate to the account settings (Me :arrow_right: Account Settings)
3. Tap "Close Account"
4. Expect that a dialog is displayed requesting the user to type their username for confirmation
5. Enter the username
6. Expect a the dialog to go into a pending state briefly (spinner), then an error dialog is displayed with a button to contact support
   * Expect that the error is tracked: `🔵 Tracked: close_account_failed, Properties: {"error_code":"active-subscriptions"}`
7. Tap "Contact Support"
8. Expect that the Help screen is displayed
   * Expect that the origin is tracked: `🔵 Tracked: support_opened, Properties: {"origin":"ACCOUNT_CLOSURE_DIALOG"}`
9. Tap back
10. Tap dismiss
11. Expect that the dialog is dismissed

#### :green_circle: Success case (disposable account)

1. Create or login with a disposable account (the account will be deleted)
2. Navigate to the account settings (Me :arrow_right: Account Settings)
3. Tap "Close Account"
4. Expect that a dialog is displayed requesting the user to type their username for confirmation
5. Enter the username
6. Expect a the dialog to go into a pending state briefly (spinner), then a success dialog is displayed with an ok button
   * Expect that the success is tracked: `🔵 Tracked: closed_account`
7. Tap "Ok"
8. Expect to be logged out and brought to the app's landing screen

## Regression Notes
1. Potential unintended areas of impact
Account settings

13. What I did to test those areas of impact (or what existing automated tests I relied on)
Manual testing

14. What automated tests I added (or what prevented me from doing so)
We do not currently have the established mocks for testing the account deletion flow, and adding that configuration would exceed the scope of this task.

PR submission checklist:


UI Changes testing checklist:
